### PR TITLE
inet: allow scopeid in uv_inet_pton

### PR DESCRIPTION
**Pull request by indutny**
We already support it in `uv_ip6_addr` anyway.

See https://github.com/joyent/node/issues/7395

### CODE REVIEW

File: src/inet.c [hunk 1]
```diff
@@ -256,7 +256,14 @@ static int inet_pton6(const char *src, unsigned char *dst) {
...
     if (ch == '.' && ((tp + sizeof(struct in_addr)) <= endp)) {
-      int err = inet_pton4(curtok, tp);
+      int err;
+
+      /* Scope id is not allowed after ipv4 address */
+      pch = strchr(curtok, '%');
+      if (pch != NULL)
+        return UV_EINVAL;
+
+      err = inet_pton4(curtok, tp);
```
Comment (saghul) [inline src/inet.c:271]: inet_pton4 will fail, I'd let it fail there.

**Comment by saghul** [comment]
uv_ip6_addr actually works with scoped addresses already. If you look at the code, it calls uv_inet_pton, so I wonder if we actually have anything to fix here. Or I'm in low-monday-mode...

File: src/inet.c (hunk 1 above)
Comment (indutny) [inline src/inet.c:271]: Oh, good point, actually.

Comment (indutny) [inline src/inet.c:271]: Fixed!

**Comment by saghul** [comment]
Ah, no, we pass a copied area which removes the scope id. I think it's ok the way we have it, I'd leave it that way.

**Comment by indutny** [comment]
So, let me describe this problem on a node level: we use `uv_inet_pton` in `net.isIP` function, the test isn't passing right now. But it'll work if manually passed in the `bind()` or `connect()` binding method.

**Comment by saghul** [comment]
I see. I propose the following: check if the system's inet_pton returns ok or not for an IP address with a scope id (I believe it doesn't). If it does, we adjust our implementation. Else, you can use uv_ip4_addr / uv_ip6_addr, or cut the original string at the '%' character. What do you think?

**Comment by indutny** [comment]
It does work:

``` C
#include <arpa/inet.h>
#include <stdio.h>

int main(int argc, char** argv) {
  struct in6_addr addr;
  int r;
  if (argc < 2) {
    fprintf(stdout, "Usage: %s <addr>\n", argv[0]);
    return 1;
  }

  r = inet_pton(AF_INET6, argv[1], &addr);
  fprintf(stdout, "Result: %d\n", r);
  return r;
}
```

``` bash
/tmp $ ./1 ::1
Result: 1
/tmp $ ./1 ::1%
Result: 1
/tmp $ ./1 "::1%ef0"
Result: 1
/tmp $ ./1 "::127.0.0.1%ef0"
Result: 1
/tmp $ ./1 "ff::ff::127.0.0.1%ef0"
Result: 0
```

**Comment by saghul** [comment]
Looks like I was wrong. At lest on OSX inet_pton seems to ignore the scope id. So half of your fix is actually correct :-)

**Comment by indutny** [comment]
Actually, initial version was better than current one, it seems :)

**Comment by saghul** [comment]
Ok, don't allow me near a computer today :-/ Sorry about that Fedor.

**Comment by indutny** [comment]
No problem, I haven't dug that deep myself. So basically it was my fault.

LGTY now?

File: test/test-ip6-addr.c [hunk 2]
```diff
@@ -97,3 +97,45 @@ TEST_IMPL(ip6_addr_link_local) {
...
+    X("abcde::1")                                                             \
+    X("fe80:0:0:0:2acf:daff:fedd:342a:5678")                                  \
+    X("fe80:0:0:0:2acf:daff:abcd:1.2.3.4")                                    \
+    X("fe80:0:0:2acf:daff:1.2.3.4.5")                                         \
+
+#define TEST_GOOD(ADDR)                                                       \
+    ASSERT(0 == uv_inet_pton(AF_INET6, ADDR, &addr));                         \
+    ASSERT(0 == uv_inet_pton(AF_INET6, ADDR "%en1", &addr));                  \
+    ASSERT(0 == uv_inet_pton(AF_INET6, ADDR "%%%%", &addr));                  \
+    ASSERT(0 == uv_inet_pton(AF_INET6, ADDR "%en1:1.2.3.4", &addr));          \
```
Comment (saghul) [inline test/test-ip6-addr.c:122]: I don't think this is a valid address.

Comment (indutny) [inline test/test-ip6-addr.c:122]: Why not? Is there any specification of the format of the interface name?

Comment (saghul) [inline test/test-ip6-addr.c:122]: Apparently doesn't give an error on OSX either. LGTM.

**Comment by indutny** [comment]
Thank you, landed in 2b21419

**Pull request merged**